rasa/core/featurizers/tracker_featurizers.py

```diff
@@ -1044,37 +1039,30 @@ def training_states_labels_and_entities(
             ):
 
                 if self.remove_duplicates:
-                    hashed = self._hash_example(tracker, states, label)
+                    hashed = self._hash_example(states, label)
                     if hashed in hashed_examples:
                         continue
                     hashed_examples.add(hashed)
 
                 # Store all positive labels associated with a training state.
-                state_hash = self._hash_example(tracker, states)
-                state_hash_to_label_list_instances[state_hash].append(label)
+                state_hash = self._hash_example(states)
+                state_hash_to_label_set[state_hash].add(label[0])
 
-                example_states.append(states)
-                example_labels.append(label)
-                example_entities.append(entities)
+                # Only add unique example states unless `remove_duplicates` is `False`.
+                if not self.remove_duplicates or state_hash not in seen_states:
+                    example_states.append(states)
+                    example_entities.append(entities)
+                    seen_states.add(state_hash)
```

Comment:
```suggestion
                state_hash = self._hash_example(states)

                # Only add unique example states unless `remove_duplicates` is `False`.
                if not self.remove_duplicates or state_hash not in state_hash_to_label_set:
                    example_states.append(states)
                    example_entities.append(entities)

                # Store all positive labels associated with a training state.
                state_hash_to_label_set[state_hash].add(label[0])
```
Sorry last comment on this section I promise :) . I just realized we could get rid of `seen_states` completely since we are already keeping the same information in `state_hash_to_label_set`.